### PR TITLE
[RFC] CLI interface for suggest API

### PR DESCRIPTION
**Pull request by himkt**
This PR introduces a new CLI command `trial suggest`, which allows users to suggest parameters in command line.

```bash
> optuna ask --storage sqlite:///sample.db --study-name tmp
[I 2021-11-20 20:23:13,355] A new study created in RDB with name: tmp
[I 2021-11-20 20:23:13,378] Asked trial 0 with parameters {}.
{"number": 0, "params": {}}
```

```bash
> optuna trial suggest int_param \
        --study-name tmp \
        --storage sqlite:///sample.db \
        --trial-number 0 \
        --distribution int \
        --low 0 \
        --high 10
{"number": 0, "params": {"int_param": 3}}
```

```bash
> optuna trial suggest float_param \
        --study-name tmp \
        --storage sqlite:///sample.db \
        --trial-number 0 \
        --distribution float \
        --low 1.0 \
        --high 10.0 \
        --log-domain
{"number": 0, "params": {"float_param": 3.463236827664237, "int_param": 3}}
```

```bash
> optuna trial suggest categorical_param \
        --study-name tmp \
        --storage sqlite:///sample.db \
        --trial-number 0 \
        --distribution categorical \
        --choices cat1 cat2 cat3
{"number": 0, "params": {"categorical_param": "cat1", "float_param": 3.463236827664237, "int_param": 3}}
```

### CODE REVIEW

**Comment by himkt** [comment]
```python
> optuna ask --storage sqlite:///sample.db --study-name sample --sampler CmaEsSampler
[I 2021-11-20 23:19:04,321] A new study created in RDB with name: sample
relative_search_space:  {}
relative_params:  {}
[I 2021-11-20 23:19:04,344] Asked trial 0 with parameters {}.
{"number": 0, "params": {}}

> optuna trial suggest i1 --study-name sample --trial-number 0 --storage sqlite:///sample.db --distribution int --high 10 --low 0
relative_search_space:  {}
relative_params:  {}
{"number": 0, "params": {"i1": 3}}

> optuna trial suggest f1 --study-name sample --trial-number 0 --storage sqlite:///sample.db --distribution float --high 1e-2 --low
relative_search_space:  {}
relative_params:  {}
{"number": 0, "params": {"f1": 0.00494407197427625, "i1": 3}}

> optuna tell --study-name sample --trial-number 0 --values 0 --storage sqlite:///sample.db
[I 2021-11-20 23:19:25,716] Told trial 0 with values [0.0] and state TrialState.COMPLETE.

> optuna ask --storage sqlite:///sample.db --study-name sample --sampler CmaEsSampler
relative_search_space:  {'f1': UniformDistribution(high=0.01, low=0.001), 'i1': IntUniformDistribution(high=10, low=0, step=1)}
relative_params:  {'f1': 0.005931305109217995, 'i1': 5}
[I 2021-11-20 23:19:33,976] Asked trial 1 with parameters {}.
{"number": 1, "params": {}}
```

Relative params is sampled.

File: optuna/trial/_trial.py [hunk 1]
```diff
@@ -63,10 +65,40 @@ def _init_relative_params(self) -> None:
...
+        if "relative_search_space" in trial_system_attrs:
+            self._load_relative_params(study, trial, trial_system_attrs)
+        else:
+            self.relative_search_space = self.study.sampler.infer_relative_search_space(
+                study, trial
+            )
+            self.relative_params = self.study.sampler.sample_relative(
+                study, trial, self.relative_search_space
+            )
+            self._write_relative_search_space_and_params()
```
Comment (himkt) [inline optuna/trial/_trial.py:78]: It introduces caches for search space and parameters. This is needed to sample parameters from a joint distribution. If this mechanism doesn't exist, `optuna trial suggest` constructs `relative_params` on each time, which means that the sampling falls into independent.

File: tests/multi_objective_tests/test_trial.py [hunk 2]
```diff
@@ -101,13 +101,14 @@ def test_system_attrs() -> None:
...
+        assert "foo" not in trial.system_attrs
         trial.set_system_attr("foo", "bar")
-        assert trial.system_attrs == {"foo": "bar"}
+        assert trial.system_attrs["foo"] == "bar"
         return [0, 0, 0]
 
     study.optimize(objective, n_trials=1)
 
-    assert study.trials[0].system_attrs == {"foo": "bar"}
+    assert study.trials[0].system_attrs["foo"] == "bar"
```
Comment (himkt) [inline tests/multi_objective_tests/test_trial.py:111]: I update this test as `trial.system_attrs` now contains relative param stuffs.
Even if we change an approach not to affect existing tests, I think this test should be independent on other attributes on `system_attrs`.

**Comment by himkt** [comment]
Now it is ready for review and discussion. 🚀

**Comment by HideakiImamura** [comment]
@toshihikoyanase Could you review this PR?

@hvy @keisuke-umezawa @ytsmiling Please leave any comments if you have.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I prefer to pass the parameter in the search space using Optuna's distribution class. I think that approach is consistent with the current ask CLI method.

File: optuna/cli.py [hunk 3]
```diff
@@ -338,6 +338,82 @@ def take_action(self, parsed_args: Namespace) -> None:
...
+        parser.add_argument("param_name")
+        parser.add_argument("--study-name", required=True)
+        parser.add_argument("--trial-number", required=True, type=int)
+        parser.add_argument("--distribution", choices=("float", "int", "categorical"))
+        parser.add_argument("--choices", nargs="+")
+        parser.add_argument("--high", type=self._int_or_float)
+        parser.add_argument("--low", type=self._int_or_float)
+        parser.add_argument("--step", type=self._int_or_float)
+        parser.add_argument("--log-domain", action="store_true", default=False)
+        parser.add_argument("--show-all-params", action="store_true", default=False)
```
Comment (HideakiImamura) [inline optuna/cli.py:363]: It looks like this argument is not used anywhere. Can we remove it?

File: optuna/cli.py [hunk 4]
```diff
@@ -338,6 +338,82 @@ def take_action(self, parsed_args: Namespace) -> None:
...
+        parser = super(_TrialSuggest, self).get_parser(prog_name)
+        parser.add_argument("param_name")
+        parser.add_argument("--study-name", required=True)
+        parser.add_argument("--trial-number", required=True, type=int)
+        parser.add_argument("--distribution", choices=("float", "int", "categorical"))
+        parser.add_argument("--choices", nargs="+")
+        parser.add_argument("--high", type=self._int_or_float)
+        parser.add_argument("--low", type=self._int_or_float)
+        parser.add_argument("--step", type=self._int_or_float)
+        parser.add_argument("--log-domain", action="store_true", default=False)
```
Comment (HideakiImamura) [inline optuna/cli.py:362]: I think we can simplify these arguments by introducing `--distribution` as a string of a distribution class like
```
optuna suggest ... --distribution UniformDistribution(0, 1) ...
```

**Review by HideakiImamura — COMMENTED** [review]
I have several comments. PTAL.

File: optuna/cli.py (hunk 4 above)
Comment (himkt) [reply optuna/cli.py:362]: Thank you for the suggestion. So, would it be nice to make `optuna trial suggest` have `--search-space` as in `optuna ask`? (if it accepts `--search-space`, we can sample multiple parameters at a time)

Comment (HideakiImamura) [reply optuna/cli.py]: Hmm... If the `optuna trial suggest` accepts multiple parameters suggestion with the `--search-space` option, the difference between the `optuna trial suggest` CLI and `Trial.suggest_*` APIs becomes larger. I prefer that the `optuna trial suggest` only accepts a single parameter with the `--distribution` option.

@toshihikoyanase Do you have any opinions on that?

File: optuna/trial/_trial.py [hunk 5]
```diff
@@ -63,10 +65,40 @@ def _init_relative_params(self) -> None:
...
+
+    def _write_relative_search_space_and_params(self) -> None:
+        relative_search_space_dict = {
+            k: distribution_to_json(d) for k, d in self.relative_search_space.items()
+        }
+        self.set_system_attr("relative_search_space", relative_search_space_dict)
+        self.set_system_attr("relative_params", self.relative_params)
+
+    def _load_relative_params(
```
Comment (HideakiImamura) [inline optuna/trial/_trial.py:87]: How about renaming the method name like `_read_relative_search_space_and_params` to make it consistent with `_write_relative_search_space_and_params`?

Comment (himkt) [reply optuna/trial/_trial.py]: It makes sense. Revised in bea69bb.

File: optuna/cli.py [hunk 6]
```diff
@@ -338,6 +338,111 @@ def take_action(self, parsed_args: Namespace) -> None:
...
+            return float(v)
+
+    def get_parser(self, prog_name: str) -> ArgumentParser:
+
+        parser = super(_TrialSuggest, self).get_parser(prog_name)
+        parser.add_argument("--study-name", required=True)
+        parser.add_argument("--trial-number", required=True, type=int)
+        parser.add_argument("--search-space", type=str, required=True)
+        parser.add_argument("--format", choices=("json", "yaml"), default="json")
+        parser.add_argument("--flatten", action="store_true")
```
Comment (toshihikoyanase) [inline optuna/cli.py:358]: Could you add `help` messages like the other subcommands?

File: tests/multi_objective_tests/test_trial.py (hunk 2 above)
Comment (toshihikoyanase) [reply tests/multi_objective_tests/test_trial.py:111]: That makes sense. Thanks!

File: optuna/trial/_trial.py [hunk 7]
```diff
@@ -63,10 +65,40 @@ def _init_relative_params(self) -> None:
...
+        trial: "optuna.trial.FrozenTrial",
+        trial_system_attrs: Dict[str, Any],
+    ) -> None:
+        relative_search_space = self.study.sampler.infer_relative_search_space(study, trial)
+        relative_search_space_cache_key = "relative_search_space"
+        relative_search_space_from_cache = {
+            k: json_to_distribution(d)
+            for k, d in trial_system_attrs[relative_search_space_cache_key].items()
+        }
+        assert relative_search_space == relative_search_space_from_cache, "search space mismatch"
```
Comment (toshihikoyanase) [inline optuna/trial/_trial.py:99]: Could you explain when this assertion fails?
And also, I'd like to know how users can fix such failure?

**Review by toshihikoyanase — COMMENTED** [review]
I'm sorry for the delayed response. I added some early comments. Please take a look.

**Comment by toshihikoyanase** [comment]
In addition, could you update the command examples in the PR description, please?

**Comment by himkt** [comment]
We decide to stop this PR after reconsidering the necessity.
I close this PR now but feel free to comment if someone want this feature!

**Pull request closed without merging**